source/agora/test/GossipProtocol.d

```diff
@@ -0,0 +1,237 @@
+        this.network.sendMessage(msg);
+    }
+}
+
+unittest {
+    import std.stdio;
+
+    TestNode n1 = new TestNode("Node1");
+    TestNode n2 = new TestNode("Node2");
+    TestNode n3 = new TestNode("Node3");
+    TestNode n4 = new TestNode("Node4");
+    TestNode n5 = new TestNode("Node5");
+
+    n1.setRemotes([n2, n3]);
+    n2.setRemotes([n4, n5]);
+
+    Hash h = sha256Of("Message No. 1");
+    n1.setMessage(h);
+
+    assert(n1.receivedMessageCount == 1);
+    assert(n2.receivedMessageCount == 1);
+    assert(n3.receivedMessageCount == 1);
+    assert(n4.receivedMessageCount == 1);
+    assert(n5.receivedMessageCount == 1);
+
+    assert(n1.lastReceivedMessage == h);
+    assert(n2.lastReceivedMessage == h);
+    assert(n3.lastReceivedMessage == h);
+    assert(n4.lastReceivedMessage == h);
+    assert(n5.lastReceivedMessage == h);
+}
+
+unittest {
+    import std.stdio;
+
+    TestNode n1 = new TestNode("Node1");
+    TestNode n2 = new TestNode("Node2");
+    TestNode n3 = new TestNode("Node3");
+    TestNode n4 = new TestNode("Node4");
+    TestNode n5 = new TestNode("Node5");
+    TestNode n6 = new TestNode("Node6");
+    TestNode n7 = new TestNode("Node7");
+
+    n1.setRemotes([n2, n3]);
+    n2.setRemotes([n4, n5]);
+    n3.setRemotes([n6, n7]);
+
+    Hash h = sha256Of("Message No. 1");
+    n1.setMessage(h);
+
+    assert(n1.receivedMessageCount == 1);
+    assert(n2.receivedMessageCount == 1);
+    assert(n3.receivedMessageCount == 1);
+    assert(n4.receivedMessageCount == 1);
+    assert(n5.receivedMessageCount == 1);
+    assert(n6.receivedMessageCount == 1);
+    assert(n7.receivedMessageCount == 1);
+
+    assert(n1.lastReceivedMessage == h);
+    assert(n2.lastReceivedMessage == h);
+    assert(n3.lastReceivedMessage == h);
+    assert(n4.lastReceivedMessage == h);
+    assert(n5.lastReceivedMessage == h);
+    assert(n6.lastReceivedMessage == h);
+    assert(n7.lastReceivedMessage == h);
+}
+
+unittest {
+    import std.stdio;
+
+    TestNode n1 = new TestNode("Node1");
+    TestNode n2 = new TestNode("Node2");
+    TestNode n3 = new TestNode("Node3");
+    TestNode n4 = new TestNode("Node4");
+    TestNode n5 = new TestNode("Node5");
+    TestNode n6 = new TestNode("Node6");
+    TestNode n7 = new TestNode("Node7");
+
+    n1.setRemotes([n2, n3]);
+    n2.setRemotes([n4, n5]);
+
+    Hash h = sha256Of("Message No. 1");
+    n1.setMessage(h);
+
+    assert(n1.receivedMessageCount == 1);
+    assert(n2.receivedMessageCount == 1);
+    assert(n3.receivedMessageCount == 1);
+    assert(n4.receivedMessageCount == 1);
+    assert(n5.receivedMessageCount == 1);
+    assert(n6.receivedMessageCount == 0);
+    assert(n7.receivedMessageCount == 0);
+
+    assert(n1.lastReceivedMessage == h);
+    assert(n2.lastReceivedMessage == h);
+    assert(n3.lastReceivedMessage == h);
+    assert(n4.lastReceivedMessage == h);
+    assert(n5.lastReceivedMessage == h);
+}
+
+unittest {
+    import std.stdio;
```

Comment:
This import can be removed, it's not used.

Reply:
Modified.

Reply:
@AndrejMitrovic 
Check Please~